### PR TITLE
Build out postgres typings

### LINKED ISSUE #60: Support postgres uuid
I'm trying to get sqlalchemy-stubs to work with a larger set of types. This would include uuids and JSON, but also encompasses a question about how to effectively create custom subtypes for columns (e.g. an Email str subtype.)

```
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer, String
from sqlalchemy.dialects.postgresql import UUID
from uuid import uuid4 as gen_uuid
from typing import NewType

Base = declarative_base()

# create NewTypes for uuids and emails so that my_email: Email = 'random string' fails.
Uuid = NewType('Uuid', str)
Email = NewType('Email', str)

def uuid4():
    return str(gen_uuid())


class UuidModel(Base):
    __abstract__ = True
    uuid = Column(UUID, primary_key=True, default=uuid4)


class User(UuidModel):
    __tablename__ = 'users'
    fico_score = Column(Integer, primary_key=True)
    first_name = Column(String)

user = User(fico_score=42, first_name='bob')
reveal_type(User.uuid)
reveal_type(user.uuid)
User.first_name.ilike('hi')
```

gives

```
script/test_model_typing.py:24: error: Revealed type is 'sqlalchemy.sql.schema.Column[Any]'
script/test_model_typing.py:25: error: Revealed type is 'Any'
```

I'd love to be able to support postgres uuids (and JSON too!) in our codebase. To accomplish this, I tried mimicking sqltypes.pyi by creating a file `sqlalchemy-stubs/dialects/postgresql/base.pyi` with contents

```
from typing import Type, NewType

from .type_api import TypeEngine as TypeEngine
from ..sql.sqltypes import _LookupExpressionAdapter


Uuid = NewType('Uuid', str)


class UUID(_LookupExpressionAdapter, TypeEngine[str]):
    __visit_name__: str = ...
    def get_dbapi_type(self, dbapi): ...
    @property
    def python_type(self) -> Type[Uuid]: ...
```

and importing it from  `sqlalchemy-stubs/dialects/postgresql/__init__.pyi` . That didn't seem to accomplish anything. I'd love some direction on the easiest way to contribute and add on to this great project.

### PR DESCRIPTION
**Pull request by bryanforbes**
Fixes #60

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
You can just git clone @bryanforbes fork (or add a remote), checkout the branch for this PR and install it (don't forget `-U`). Then check if this PR fixes your problem (most likely it does, with this PR `reveal_type(user.uuid)` should be `str`).

**Comment by ilevkivskyi** [comment]
(I will try to review this PR by the end of the week, but I am not sure it will get into 0.1 release on PyPI.)

**Comment by kespindler** [comment]
Yeah, we'll try this soon and report back progress.

Have you put any thought into making the inferred type extensible to support the NewTypes I mention in #60 ? (Specifically, creating a `Uuid = NewType('Uuid', str)` so that `my_uuid: Uuid = 'random string'` fails typechecking without an explicit cast to Uuid.) The goal would for the class attr to typecheck to `Revealed type is 'sqlalchemy.sql.schema.Column[Uuid]'` and object attr to `Revealed type is 'Any'`. This is also super useful for Email, E164 Phone Numbers, and a slew of other types that have meaningful structure, but are encoded in `String(...)` fields.

**Comment by ilevkivskyi** [comment]
If you want the support for `NewType`s you will need to define your own (trivial) `TypeDecorator` and make it generic (requires some `if TYPE_CHECKING` tricks because nothing in SQLAlchemy is generic at runtime). I don't want to make `Enum` itself generic since this will cause false positives for people who don't use `NewType`s.

Also this is off-topic for this PR, please move the discussion to the issue.

File: sqlalchemy-stubs/dialects/postgresql/ranges.pyi [hunk 1]
```diff
@@ -0,0 +1,36 @@
+from ... import types as sqltypes
+from ...sql import type_api
+from typing import Any
+
+class RangeOperators: ...
+#     class comparator_factory(type_api.TypeEngine.Comparator):
```
[the remark this answers is not in the document]
Comment (ilevkivskyi) [reply sqlalchemy-stubs/dialects/postgresql/ranges.pyi]: This should be fixed on mypy master, let's check if tests pass.

File: sqlalchemy-stubs/dialects/postgresql/__init__.pyi [hunk 2]
```diff
@@ -1,5 +1,22 @@
 from typing import Any as _AnyType
 
+from .base import \
+    INTEGER as INTEGER, BIGINT as BIGINT, SMALLINT as SMALLINT, VARCHAR as VARCHAR, CHAR as CHAR, \
+    TEXT as TEXT, NUMERIC as NUMERIC, FLOAT as FLOAT, REAL as REAL, INET as INET, CIDR as CIDR, \
+    UUID as UUID, BIT as BIT, MACADDR as MACADDR, MONEY as MONEY, OID as OID, REGCLASS as REGCLASS, \
+    DOUBLE_PRECISION as DOUBLE_PRECISION, TIMESTAMP as TIMESTAMP, TIME as TIME, DATE as DATE, \
+    BYTEA as BYTEA, BOOLEAN as BOOLEAN, INTERVAL as INTERVAL, ENUM as ENUM, TSVECTOR as TSVECTOR, \
+    DropEnumType as DropEnumType, CreateEnumType as CreateEnumType
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/dialects/postgresql/__init__.pyi]: TBH I dom't like the `\` style. Could you please update all long imports in this file and below to be like:
```python
from pack.module import (
    ClassOne, ClassTwo, ClassThree,
    OtherStuff, Whatever
)

File: sqlalchemy-stubs/dialects/postgresql/ext.pyi [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+class ExcludeConstraint(ColumnCollectionConstraint):
+    __visit_name__: str = ...
+    where: Optional[visitors.Visitable] = ...
+    operators: Dict[str, Any] = ...
+    using: str = ...
+    def __init__(self, *elements: Any, name: str = ..., deferrable: bool = ..., initially: str = ..., using: str = ...,
+                 where: Optional[Union[str, bool, visitors.Visitable]] = ...) -> None: ...
+    def copy(self: _EC, **kw: Any) -> _EC: ...
+
+def array_agg(*arg: Any, **kw: Any) -> functions.array_agg[_T]: ...
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/dialects/postgresql/ext.pyi]: Why the type variable only appears in the return type? I think this doesn't make much sense.

File: sqlalchemy-stubs/dialects/postgresql/hstore.pyi [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+class _HStoreValsFunction(sqlfunc.GenericFunction[List[Text]]):
+    type: ARRAY[Text] = ...
+    name: str = ...
+
+class _HStoreArrayFunction(sqlfunc.GenericFunction[List[Text]]):
+    type: ARRAY[Text] = ...
+    name: str = ...
+
+class _HStoreMatrixFunction(sqlfunc.GenericFunction[List[Text]]):
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/dialects/postgresql/hstore.pyi]: Just to double-check: do we really need al these underscored (private?) classes?

File: sqlalchemy-stubs/sql/type_api.pyi [hunk 5]
```diff
@@ -109,8 +110,9 @@ class Variant(TypeDecorator[_T]):
     def coerce_compared_value(self: _VT, operator: Any, value: Any) -> Union[_VT, TypeEngine[_T]]: ...
     def load_dialect_impl(self, dialect: Dialect) -> TypeEngine[Any]: ...
     def with_variant(self, type_: TypeEngine[_U], dialect_name: str) -> Variant[_U]: ...
+    # Signature of comparator factory incompatible with supertype TypeEngine
     @property
-    def comparator_factory(self) -> Type[Any]: ...
+    def comparator_factory(self) -> Type[Any]: ...  # type: ignore
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/type_api.pyi]: There is some inconsistency in how `comparator_factory` is defined in various classes, sometimes it is a read-only property, but sometimes it is a writeable attribute. Maybe these properties are actually settable so we can replace them with just normal attributes?

File: sqlalchemy-stubs/dialects/postgresql/ext.pyi (hunk 3 above)
Comment (bryanforbes) [reply sqlalchemy-stubs/dialects/postgresql/ext.pyi]: Yeah, this typing is wrong. I'll fix it.

File: sqlalchemy-stubs/dialects/postgresql/hstore.pyi (hunk 4 above)
Comment (bryanforbes) [reply sqlalchemy-stubs/dialects/postgresql/hstore.pyi]: I didn't fill out all of the return types for the comparator functions which is where these are used. I'll do that now.

File: sqlalchemy-stubs/dialects/postgresql/ranges.pyi (hunk 1 above)
[the remark this answers is not in the document]
Comment (bryanforbes) [reply sqlalchemy-stubs/dialects/postgresql/ranges.pyi]: This does seem to be fixed by mypy master. Thanks!

File: sqlalchemy-stubs/sql/type_api.pyi (hunk 5 above)
Comment (bryanforbes) [reply sqlalchemy-stubs/sql/type_api.pyi]: For `Variant` and `TypeDecorator`, `comparator_factory` is calculated from the implementation, so it has to be a read-only property. We could type it as read/write, but I feel wrong doing that.

File: sqlalchemy-stubs/dialects/postgresql/__init__.pyi [hunk 6]
```diff
@@ -1,5 +1,25 @@
...
+    array_agg as array_agg
+)
+from .dml import insert as insert, Insert as Insert
+
+from .ranges import (
+    INT4RANGE as INT4RANGE, INT8RANGE as INT8RANGE, NUMRANGE as NUMRANGE, DATERANGE as DATERANGE,
+    TSRANGE as TSRANGE, TSTZRANGE as TSTZRANGE
+)
 
 def __getattr__(attr: str) -> _AnyType: ...
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/dialects/postgresql/__init__.pyi]: Do we still need this `__getattr__()` or the stubs are now complete?

File: sqlalchemy-stubs/sql/functions.pyi [hunk 7]
```diff
@@ -74,7 +75,22 @@ class GenericFunction(Function[_T], metaclass=_GenericMeta):
     packagenames: Any = ...
     clause_expr: Grouping[Any] = ...
     type: TypeEngine[_T] = ...
-    def __init__(self, *args: Any, **kwargs: Any) -> None: ...
+    @overload
+    def __init__(self, *args: ColumnElement[Any], bind: Optional[Union[Engine, Connection]] = ...,
+                 type_: Type[TypeEngine[_T]]) -> None: ...
```
Comment (ilevkivskyi) [inline sqlalchemy-stubs/sql/functions.pyi]: This now causes an error on `functions.concat(users.c.user, "some.suffix")`. But it works at runtime.

Comment (ilevkivskyi) [reply sqlalchemy-stubs/sql/functions.pyi]: Since this is the only remaining small issue, I am going to merge this now, you can fix it in a separate PR.

**Pull request merged**